src/communication/network_interface_sr.c

```diff
@@ -7268,6 +7269,208 @@ slogtb_dump_trantable (THREAD_ENTRY * thread_p, unsigned int rid, char *request,
+
+	  ptr = or_pack_int (ptr, c_info->num_subcomp);
+
+	  for (int j = 0; j < c_info->num_subcomp; j++)
+	    {
+	      s_info = &(c_info->subcomp_info[j]);
+	      ptr = or_pack_string (ptr, s_info->name);
+	      ptr = or_pack_int64 (ptr, s_info->cur_stat);
+	    }
+	}
+
+      /* 4) deallocate memory */
+      for (int i = 0; i < info.num_comp; i++)
+	{
+	  free (info.comp_info[i].subcomp_info);
+	}
+      free (info.comp_info);
+    }
+
+  if (error != NO_ERROR)
+    {
+      ptr = or_pack_int (reply, 0);
+      ptr = or_pack_int (ptr, error);
+      css_send_data_to_client (thread_p->conn_entry, rid, reply, OR_ALIGNED_BUF_SIZE (a_reply));
+      free_and_init (buffer);
+    }
+  else
+    {
+      ptr = or_pack_int (reply, size);
+      ptr = or_pack_int (ptr, error);
+      css_send_data_to_client (thread_p->conn_entry, rid, reply, OR_ALIGNED_BUF_SIZE (a_reply));
+      css_send_data_to_client (thread_p->conn_entry, rid, buffer, size);
+      free_and_init (buffer);
+    }
+}
+
+
+/*
+ * smmon_get_tran_info -
+ *
+ * return:
+ *
+ *   rid(in):
+ *   request(in):
+ *   reqlen(in):
+ */
+void
+smmon_get_tran_info (THREAD_ENTRY * thread_p, unsigned int rid, char *request, int reqlen)
+{
+  char *buffer, *ptr;
+  int size = 0;
+  OR_ALIGNED_BUF (OR_INT_SIZE + OR_INT_SIZE) a_reply;
+  char *reply = OR_ALIGNED_BUF_START (a_reply);
+  int error = NO_ERROR;
+  MMON_TRAN_INFO info;
+  int tran_count = 0;
+
+  ptr = or_unpack_int (request, &tran_count);
+
+  error = mmon_aggregate_tran_info (info, tran_count);
+
+  if (error == NO_ERROR)
+    {
+      /* before send information 
+       * 1) check size
+       * 2) allocate buffer
+       * 3) packing information 
+       * 4) deallocate MMON_TRAN_INFO */
+
+      /* 1) check size 
+       *    - calculate buffer size of packed values
+       *    - server info -> transaction info */
+      size += or_packed_string_length (info.server_info.name, NULL);
+      size += OR_INT64_SIZE;
+
+      size += OR_INT_SIZE;
+
+      for (int i = 0; i < info.num_tran; i++)
+	{
+	  size += OR_INT_SIZE + OR_INT64_SIZE;
+	}
+
+      /* 2) allocate buffer */
+      buffer = (char *) malloc (size);
+
+      /* 3) packing information
+       *    - packing information with sequence of
+       *      server info -> transaction info */
+      ptr = or_pack_string (buffer, info.server_info.name);
+      ptr = or_pack_int64 (ptr, info.server_info.total_mem_usage);
+
+      ptr = or_pack_int (ptr, info.num_tran);
+
+      for (int i = 0; i < info.num_tran; i++)
+	{
+	  ptr = or_pack_int (ptr, info.tran_stat[i].tranid);
+	  ptr = or_pack_int64 (ptr, info.tran_stat[i].cur_stat);
+	}
+
+      /* 4) deallocate memory */
+      free (info.tran_stat);
+    }
+
+  if (error != NO_ERROR)
+    {
+      ptr = or_pack_int (reply, 0);
+      ptr = or_pack_int (ptr, error);
+      css_send_data_to_client (thread_p->conn_entry, rid, reply, OR_ALIGNED_BUF_SIZE (a_reply));
+      free_and_init (buffer);
```

Comment:
free_and_init () is duplicated. I think all the `free ()` jobs can be done at the end of the function